hoverkite-firmware/src/protocol.rs

```diff
@@ -96,157 +97,68 @@ pub fn process_command(
     target_position: &mut Option<i64>,
     spring_constant: &mut i64,
 ) -> bool {
-    if command.len() < 1 {
-        return false;
-    }
-
-    match command[0] {
-        b'F' => {
-            if command.len() < 2 {
-                return false;
-            }
-            let forward_length = command[1];
-            if command.len() < forward_length as usize + 2 {
-                return false;
-            }
-            forward_command(hoverboard, &command[2..]);
-        }
-        b'l' => {
-            if command.len() < 2 {
-                return false;
-            }
-            match command[1] {
-                b'0' => {
-                    log!(hoverboard.response_tx(), "LED off");
-                    hoverboard.leds.side.set_low().unwrap()
-                }
-                b'1' => {
-                    log!(hoverboard.response_tx(), "LED on");
-                    hoverboard.leds.side.set_high().unwrap()
-                }
-                _ => log!(hoverboard.response_tx(), "LED unrecognised"),
-            }
-        }
-        b'o' => {
-            if command.len() < 2 {
-                return false;
-            }
-            match command[1] {
-                b'0' => {
-                    log!(hoverboard.response_tx(), "orange off");
-                    hoverboard.leds.orange.set_low().unwrap()
-                }
-                b'1' => {
-                    log!(hoverboard.response_tx(), "orange on");
-                    hoverboard.leds.orange.set_high().unwrap()
+    let message = match Message::parse(command) {
+        Ok(message) => message,
+        Err(nb::Error::WouldBlock) => return false,
+        Err(err) => {
+            log!(
+                hoverboard.response_tx(),
+                "Unrecognised command {} or problem {:?}",
+                command[0],
+                err
+            );
+            // make sure the buffer progresses here
+            return true;
+        }
+    };
+
+    match message {
+        Message::SecondaryCommand(sc) => {
+            forward_command(hoverboard, &sc.0);
+        }
+        Message::Command(c) => match c {
+            Command::ReportBattery => {
+                let readings = hoverboard.adc_readings();
```

Comment:
sorry: this section ended up getting re-indented. Use w=1 as a url parameter to ignore whitespace for the below section if it helps (https://github.com/hoverkite/hoverkite/pull/11/files?diff=unified&w=1)